### ISSUE TITLE
Create generic S3 protocol handler in `UriSource` and `UriSink`

### ISSUE THREAD

**Issue by senderista** [uwescience/myria#797]
We should have a generic S3 protocol handler for import/export, using the Hadoop S3 `FileSystem` implementation (the [`s3a`](http://hadoop.apache.org/docs/r2.7.2/hadoop-aws/tools/hadoop-aws/index.html) one to be precise, since it's built on the AWS SDK, and hence among other things allows us to handle credentials securely). This is necessary for exporting any data to S3, and for importing any S3 data that isn't publicly readable (and thus accessible through HTTP). Specific optimizations for parallel S3 import/export will be addressed in separate issues.

This will require us to obtain credentials for calling into the AWS SDK. The most secure way to handle credentials is to require them to be present on an EC2 instance via an IAM role used to launch the instance: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html. This would require us to add support for IAM roles to the deployment script, which shouldn't be difficult, but would also require the AWS identity running the deployment script to have the `PassRole` permission for the IAM role they want to pass to their cluster. Alternatively, we could propagate the user's AWS credentials directly to the cluster by getting them from `boto`, passing them to our Ansible playbook, and writing them as files on the remote machine for use by [`envdir`](http://thedjbway.b0llix.net/daemontools/envdir.html), then wrapping all launch commands for Myria executables in an `envdir` invocation with the location of the credentials. This isn't as secure as launching under an IAM role, but should be secure enough, provided we create a dedicated unprivileged user for the `myria` service, with exclusive read permissions on the `envdir` files containing AWS credentials. I favor the first approach (launching the cluster under an IAM role), since it's much simpler and more secure, but we have to evaluate the inconvenience to users who may have difficulty obtaining `PassRole` privileges. In any case, I think the first approach should be implemented first.

It [appears](https://issues.apache.org/jira/browse/HADOOP-11571) that the `s3a` implementation was significantly improved in Hadoop 2.7, so we should probably upgrade to the latest 2.7 version before deploying this feature (unless it's incompatible with REEF, which is [still](https://github.com/markusweimer/incubator-reef/blob/master/pom.xml#L51) at Hadoop 2.6.0).

**Comment by senderista** [comment]
@jingjingwang would appreciate your feedback.

**Comment by BrandonHaynes** [comment]
Since the UriSink class is now going to contain separate, dedicated logic for HTTP(s), HDFS and now S3, it's probably worth having it delegate to specific strategies by scheme.

**Comment by senderista** [comment]
Depends on whether we can just pass the URI to `FileSystem.get()`. If we do have to implement specialized logic for each scheme, then I agree, but hopefully for now we can get away with the above for HDFS and S3 URIs (except we might want to munge the `s3` scheme to `s3a` if we're using the new S3A filesystem).

What would be the application for exporting to an HTTP URI? Would this be a `PUT` or `POST` call, and we would just hope the server supports those verbs? Even if the server did support them, they would likely be expecting particular headers etc., so I don't see how we could generically support export to HTTP.

**Comment by jingjingwang** [comment]
Seems the only possible disadvantage of the first approach is the difficulty of obtaining `PassRole`? I'm not very familiar with `PassRole`  but in my imagination the overhead should be negligible since the user needs to deal with AWS credentials anyway, and using IAM is recommended by Amazon. 

I hope Hadoop 2.7 is compatible with REEF :) But if not then we may have to postpone the upgrade...

**Comment by jingjingwang** [comment]
I think exporting to an HTTP URI is more like something that we can have easily and can reserve for future usage. For example, a data visualization service? Although we don't have an immediate use case for now.

**Comment by senderista** [comment]
I've added `PassRole` support to the deploy script/playbook and upgraded Hadoop to 2.7. I created a role `myria-cluster` under the Myria account which has full access to our S3 resources and no other permissions. That's enough for now, but probably won't be later. You can try this now: `myria-deploy --role myria-cluster`. Everything seems to work fine in my testing so far (I installed awscli on the coordinator with zero config so it was forced to use the IAM role credentials and tried some basic S3 operations).

**Comment by senderista** [comment]
Raco support for this feature will be implemented in https://github.com/uwescience/raco/issues/496

**Comment by senderista** [comment]
This is blocking on https://github.com/uwescience/myria/issues/802

**Comment by senderista** [comment]
@jortiz16 I didn't realize until I was testing the `s3-uri-reef` branch that S3 import wasn't part of this issue. I don't think it's worth opening a separate issue for the `UriSource` changes/new tests for S3 support because the changes are so similar. So I'd like to add `UriSource` support for the S3 protocol to this issue and associated PRs. (The linked Raco issue already covers `IMPORT`/`EXPORT` implementation.) Let me know if you feel differently.

**Comment by jortiz16** [comment]
This is ok. Should I then keep building on s3-uri-reef? I noticed you moved my commits there. On Friday, I tried pushing commits to add-s3-UriSink but the tests still did not work on Travis when I tried to export to S3 --  I would still get the "No FileSystem for scheme: s3a" as seen [here](https://travis-ci.org/uwescience/myria/jobs/108947929). I don't have enough understanding of Travis to get a sense for what the problem is. Or is it that we don't have the correct Hadoop jars/dependencies on the Myria master branch for S3 to work?

**Comment by senderista** [comment]
I cherry-picked your commits from `add-s3-UriSink` into `s3-uri-reef` just so I could do my own testing on EC2 (I wanted to try `load()` with an S3 URL, which worked after I had fixed the YARN classpath to include the JAR containing the `s3a` code, and also made an analogous fix to `DataSource` to handle S3 URLs). I've pushed a commit to `add-s3-UriSink` which *should* fix your issue running S3 tests on Travis, but I haven't tested that yet. (I *have* tested S3 writes from Travis using `boto` with the encrypted credentials I added to `.travis.yml`.)

**Comment by senderista** [comment]
@jortiz16 testing in Travis is currently blocked by https://github.com/uwescience/myria/issues/809

**Comment by senderista** [comment]
Fixed in https://github.com/uwescience/myria/issues/811
